### PR TITLE
tlv: fuzz test encoding/decoding

### PR DESCRIPTION
**Pull request by morehouse**
Based on https://github.com/lightningnetwork/lnd/pull/4641.

### CODE REVIEW

File: tlv/fuzz_test.go [hunk 1]
```diff
@@ -96,3 +96,42 @@ func FuzzVarBytes(f *testing.F) {
 		harness(t, data, EVarBytes, DVarBytes, &val, uint64(len(data)))
 	})
 }
+
+// bigSizeHarness works the same as harness, except that it compares decoded
+// values instead of encoded values. We do this because DBigSize may leave some
+// bytes unparsed from data, causing the encoded data to be shorter than the
+// original.
+func bigSizeHarness(t *testing.T, data []byte, val1, val2 interface{}) {
+	if len(data) > 9 {
```
Comment (yyforyongyu) [inline tlv/fuzz_test.go:105]: I'm starting to wonder how often this is hit and whether it deteriorates the efficacy of our test?

File: tlv/fuzz_test.go [hunk 2]
```diff
@@ -96,3 +96,42 @@ func FuzzVarBytes(f *testing.F) {
...
+		return
+	}
+
+	r := bytes.NewReader(data)
+
+	var buf [8]byte
+	if err := DBigSize(r, val1, &buf, 9); err != nil {
+		return
```
Comment (yyforyongyu) [inline tlv/fuzz_test.go:113]: why can we return here instead of assertion?

File: tlv/fuzz_test.go [hunk 3]
```diff
@@ -135,3 +135,36 @@ func FuzzBigSize64(f *testing.F) {
...
+		}
+
+		var val uint32
+		harness(t, data, ETUint32, DTUint32, &val, uint64(decodeLen))
+	})
+}
+
+func FuzzTUint64(f *testing.F) {
+	f.Fuzz(func(t *testing.T, data []byte, decodeLen uint8) {
+		if decodeLen > 8 {
```
Comment (yyforyongyu) [inline tlv/fuzz_test.go:163]: what do you say if we just iterate from 0 to 7 here?

File: tlv/fuzz_test.go [hunk 4]
```diff
@@ -168,3 +168,111 @@ func FuzzTUint64(f *testing.F) {
...
+			u8   uint8
+			u16  uint16
+			u32  uint32
+			u64  uint64
+			b32  [32]byte
+			b33  [33]byte
+			b64  [64]byte
+			pk   *btcec.PublicKey
+			b    []byte
+			bs32 uint32
```
Comment (yyforyongyu) [inline tlv/fuzz_test.go:214]: curious about what `bs` and `tu` stand for?

File: tlv/fuzz_test.go [hunk 5]
```diff
@@ -168,3 +168,111 @@ func FuzzTUint64(f *testing.F) {
...
 }
+
+// encodeParsedTypes re-encodes TLVs decoded from a stream, using the
+// parsedTypes and decodedRecords produced during decoding.
+func encodeParsedTypes(t *testing.T, parsedTypes TypeMap,
+	decodedRecords []Record) []byte {
+
+	var encodeRecords []Record
+	for typ, val := range parsedTypes {
+		if typ < Type(len(decodedRecords)) {
```
Comment (yyforyongyu) [inline tlv/fuzz_test.go:179]: would be helpful if we add some docs explaining why this check.

**Review by yyforyongyu — COMMENTED** [review]
Sorry for the late review. Left some questions, and this PR needs a rebase.

File: tlv/fuzz_test.go (hunk 4 above)
Comment (morehouse) [reply tlv/fuzz_test.go:222]: BigSize and Truncated Uint.

File: tlv/fuzz_test.go (hunk 5 above)
Comment (morehouse) [reply tlv/fuzz_test.go:185]: Done in https://github.com/lightningnetwork/lnd/pull/7889/commits/5bb4df424c2a2a926b881449c9585fed6174e6da.

**Review by morehouse — COMMENTED** [review]
Rebased and updated.

**Review by Crypt-iQ — COMMENTED** [review]
looks good, will test locally

**Review by yyforyongyu — APPROVED** [review]
LGTM🙏

**Review by Crypt-iQ — APPROVED** [review]

**Pull request merged**